### ISSUE TITLE
Image processing plugin

### ISSUE THREAD

**Issue by kevinlang** [lumeland/lume#42]
It would be nice to have a plugin that can take a source image and output various optimized formats, leveraging `srcset` and the like to enable responsive images.

As far as I can tell, this is blocked by https://github.com/lovell/sharp/issues/2583 which is blocked by upstream Deno support.

**Comment by oscarotero** [comment]
I've used sharp in other node projects and it's an incredible library, and a Deno version would allow us to create plugin similar to Eleventy (https://www.11ty.dev/docs/plugins/image/)

For now, the only way is a script on "afterBuild" event, for example using imagick:

```js
site.addEventListener("afterBuild", "convert -resize 800x800 _site/*.jpg");
```

But definitelly, if sharp is ported to Deno, we will have a image processing plugin.

**Comment by oscarotero** [comment]
I have been playing the with deno-image (https://github.com/MariusVatasoiu/deno-image) package and I think it can be used for a first version of this plugin, until sharp is available for Deno.

It's a very limited package, because it supports jpg and png formats and only can resize them, but it could be useful in some cases. I couldn't find any other Deno package with better results (https://deno.land/x?query=image)

[ImageScript](https://deno.land/x/imagescript@1.1.19) looks great but it only supports nearest neighbor scaling mode, so the output has poor quality.

The WIP plugin is here: https://github.com/lumeland/lume/blob/master/plugins/image.js and the way to use it is:

```js
site.use(image({
  resize: {
    "/img/avatars": {
      "-small": { width: 100, height: 100 },
      "-medium": { width: 500, height: 500 },
      "-big": { width: 1000, height: 1000 }
    },
    "/img/pictures/*.jpg": {
      "-mobile": { width: 500, height: 500 },
      "-tablet": { width: 1000, height: 1000 },
      "-desktop": { width: 1500, height: 1500 }
    }
  }
}));
```

The plugin take all images of `/img/avatars` folder and create three versions appending the suffix `-small`, `-medium` and `-big` to the new versions. You can use also glob patterns, as you can see in the second example `/img/pictures/*.jpg`.

**Comment by oscarotero** [comment]
ImageScript has the `rewrite` branch that looks awesome: https://github.com/matmen/ImageScript/tree/rewrite
It includes a lot of improvements, like Typescript and more interpolation modes, so it's a good alternative to sharp.

**Comment by oscarotero** [comment]
Squoosh is interesting. I never worked with WASM but looks like a Deno port should be feasible. If anyone want to work on this, that would be awesome.

BTW, @enjikaka I like you approach of using the HTML as the source to get and convert the images. It's similar to other plugins like [inline](https://github.com/lumeland/lume/blob/master/plugins/inline.js). I'll change this plugin for something more similar to yours.

**Comment by enjikaka** [comment]
If it's keeps raining outside I might give the Deno/WASM/Squoosh combo a try; otherwise garden word has higher priority here. 👨🏻‍🌾 

@oscarotero Glad you like it. I haven't thought it through completely either, so the HTML-approach might not be sustainable in the long run. But it worked for now at least. It would be a lot cleaner code with proper path handling and DOM parsing like in the inline plugin!

**Comment by oscarotero** [comment]
Garden has always priority over programing, agreed.

And about the html-approach, yes, I'm refering to parse and use DOM API to search the images. I'm thinking on something like this:

```html
<img src="image.jpg" resize="300x200">
```

**Comment by oscarotero** [comment]
More ideas: https://rbuchberger.github.io/jekyll_picture_tag/

**Comment by oscarotero** [comment]
I am experimenting with the new `imagick` plugin: https://github.com/lumeland/experimental-plugins/tree/main/imagick
It allows to process images using ImageMagick and generate different versions.

It uses `_data` files to assign different processing functions. Example:
- Modify the images in this directory, apply a resize and blur: https://github.com/lumeland/experimental-plugins/blob/main/imagick/demo/img/_data.ts
- Modify the images in this directory, generate three images with different sizes: https://github.com/lumeland/experimental-plugins/blob/main/imagick/demo/avatars/_data.ts

**Comment by enjikaka** [comment]
@oscarotero  Cool! I don't understand from where the method in your _data.ts file is called?

On a side note I put my squoosh experiment in a repo of its own: https://github.com/enjikaka/lume_plugin_squoosh

**Comment by oscarotero** [comment]
@enjikaka 
- It uses `loadAssets` to load the image files, so they are pages that can be (pre)processed.
- The processor looks for the `imagick` value [here](https://github.com/lumeland/experimental-plugins/blob/main/imagick/imagick.ts#L36)
- If this value doesn't exist, do nothing. If exists, execute it.
- It returns an object with filenames as keys and transform functions as values. For example:
  ```js
  {
    "/filename.jpg": (img) => img.resize(200, 200),
    "/filename-small.jpg": (img) => img.resize(100, 100),
    "/filename-big.jpg": (img) => img.resize(300, 300),
  }
  ```
- It generates new files for every entry of this object, using the ImageMagick API to transform the original image.
- In the demo, I've used a `_data.ts` file to apply the same function to all files of a directory. This allows to assign different functions to files in different directories.
- Due the function is assigned like any other regular `data` variable, you can use _data files or use other processor to set this value automatically. For example, to resize all jpg files, you can assign the function with this processor:
  ```js
  site.process([".jpg", (file) => {
      file.data.imagick = function (url) {
        return {
          [url]: (img) => img.resize(150, 150)
        }
      };
  });
- This opens the door to create other plugins that can generate the imagick functions automatically, for example, by scanning the images in the HTML and parsing special attributes.

**Comment by oscarotero** [comment]
Imagick plugin released: https://lume.land/plugins/imagick/
